File: app/src/main/java/org/thoughtcrime/securesms/gcm/FcmUtil.java
package org.thoughtcrime.securesms.gcm;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.WorkerThread;

import com.google.android.gms.tasks.Tasks;
import com.google.firebase.FirebaseApp;
import com.google.firebase.messaging.FirebaseMessaging;

import org.signal.core.util.logging.Log;

import java.util.Optional;
import java.util.concurrent.ExecutionException;

public final class FcmUtil {

  private static final String TAG = Log.tag(FcmUtil.class);

  /**
   * Retrieves the current FCM token. If one isn't available, it'll be generated.
   */
  @WorkerThread
  public static Optional<String> getToken(Context context) {
    String token = null;

    // Must be called manually if running outside of main process
    FirebaseApp.initializeApp(context);

    try {
      token = Tasks.await(FirebaseMessaging.getInstance().getToken());
    } catch (InterruptedException e) {
      Log.w(TAG, "Was interrupted while waiting for the token.");
    } catch (ExecutionException e) {
      Log.w(TAG, "Failed to get the token.", e.getCause());
    }

    return Optional.ofNullable(TextUtils.isEmpty(token) ? null : token);
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/gcm/FcmUtil.kt
package org.thoughtcrime.securesms.gcm;

import androidx.annotation.WorkerThread

import com.google.android.gms.tasks.Tasks
import com.google.firebase.messaging.FirebaseMessaging

import org.signal.core.util.logging.Log
import org.whispersystems.libsignal.util.guava.Optional

import java.io.IOException
import java.util.concurrent.ExecutionException

object FcmUtil {

  private val TAG: String = Log.tag(FcmUtil.javaClass)

  /**
   * Retrieves the current FCM token. If one isn't available, it'll be generated.
   */
  @JvmStatic
  @WorkerThread
  @Throws(FCMDisabledException::class)
  fun getToken(): Optional<String> {
    var token: String? = null

    try {
      token = Tasks.await(FirebaseMessaging.getInstance().token)
    } catch (e: InterruptedException) {
      Log.w(TAG, "Was interrupted while waiting for the token.")
    } catch (e: IOException) { // IGNORE THIS WARNING
      throw FCMDisabledException()
    } catch (e: ExecutionException) {
      Log.w(TAG, "Failed to get the token.", e.cause)
    }


    return Optional.fromNullable(token?.takeIf { token.isNotEmpty() })
  }

  /**
   * MicroG can disable FCM.
   *
   * They do this by throwing an [IOException] with a message of SERVICE_NOT_AVAILABLE.
   *
   * To respond properly, set [AccountValues.fcmEnabled] to false and not use FCM again.
   */
  class FCMDisabledException : Exception("MicroG disabled this functionality")
}
